### PR TITLE
[Issue 6024][pulsar_storm] Support alternate output streams in PulsarSpout

### LINKED ISSUE #6024: PulsarSpout emits only to default stream
**Is your feature request related to a problem? Please describe.**
We have an application, currently using another streaming ingestion method, that we'd like to replace with either Kafka or Pulsar. Either would be a candidate, _except_ the spout needs to emit to different output streams (to be processed by different bolts) based on the content of individual messages.

The Kafka spout supports this. Topologies implement `RecordTranslator`, an interface that returns the output stream and the `Values` for the tuple.

The equivalent in `PulsarSpout` is `MessageToValuesMapper`, but that returns only the `Values`. There's no way to specify an alternate stream; everything goes to "default."

**Describe the solution you'd like**
Simply put, a way for `MessageToValuesMapper` to specify the output stream for a tuple. (It already has a `declareOutputFields()` method so can declare the appropriate output streams.)

Three possible solutions, all workable. The first avoids any API change but requires an extra check
in `PulsarSpout` on emit. The other two would require backwards-compatible changes to `MessageToValueMapper`. I prefer the first.

1. Define `PulsarTuple` as an extension of `Values` so `MessageToValuesMapper` can return it without changing the API. `PulsarSpout` checks for `PulsarTuple` and emits on the specified stream if found. (_This is the alternative in the pull request._)
1. Change `PulsarSpout` to use a new method in `MessageToValuesMapper` that returns a `PulsarTuple` interface similar to Kafka's `KafkaTuple`. It would return the `Values` _and_ output stream. A default implementation of the new method would just call the existing method (which probably should be deprecated) and return a `PulsarTuple` with stream "default."
1. Add a new method to `MessageToValuesMapper`: `String outputStream(Message msg, Values values)` to return the output stream given the original `Message` and extracted `Values`. The default implementation would just return "default."

**Describe alternatives you've considered**
All of these would require `collector` to be not `private`, and probably other things. None are very satisfying. All would require an application to extend `PulsarSpout`, which can cause compatibility issues with new versions.
1. Change the existing `collector.emit()` calls with a call to a non-private `doEmit(Message msg, Values values)` call. Applications could override that and figure out the output stream as needed. 
1. Make `mapToValueAndEmit()` at least protected so it can be overridden.
1. Override nextTuple() and replicate most of the logic except the `emit()` calls.

Or, just use Kafka, which is what we'd actually do.

**Additional context**
One application receives Avro-encoded events with a header that identifies the schema type and version. (Also some other information not relevant here.) The schema type determines the output stream.

Another application has a fixed Avro schema with two fields, an event type and a serialized protobuf. The event type determines the output stream.

Changing those formats isn't an option.

### PR DESCRIPTION
**Pull request by dawillcox**
Fixes #6024

### Motivation
This is all described in detail in https://github.com/apache/pulsar/issues/6024, but in short, an insurmountable obstacle to using Pulsar in our storm topology is the fact that `PulsarSpout` only emits to the "default" stream. In our environment, we need to emit on different streams based on the content of each received message. This change extends `PulsarSpout` to recognize a `Values` extension that specifies an alternate output stream, and uses that stream when given.

### Modifications
A new `PulsarTuple` class is added. It extends `Values` and adds a method to return the output stream.

When emitting a tuple after calling `toValues(msg)`, `PulsarSpout` checks if the returned `Values` is a `PulsarTuple`. If so, it emits to the designated stream, otherwise it emits as before.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - A test case was added to `PulsarSpoutTest` to verify that when `PulsarTuple` us used, the tuple is emitted in the specified stream.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes?) Existing APIs don't change, but a new class is added that carries additional information.
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/MessageToValuesMapper.java [hunk 1]
```diff
@@ -18,27 +18,29 @@
...
     /**
      * Convert {@link org.apache.pulsar.client.api.Message} to tuple values.
+     * Note: If the returned Values is an instance of
+     * {@link org.apache.pulsar.storm.PulsarTuple}, it specifies which
+     * output stream to emit to.
      *
      * @param msg
      * @return
      */
-    public Values toValues(Message msg);
```
Comment (dawillcox) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/MessageToValuesMapper.java]: Specifying `public` here is redundant. Interface methods _must_ be public.

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java [hunk 2]
```diff
@@ -18,14 +18,13 @@
...
 
-import java.util.Map;
-import java.util.Queue;
-import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.TimeUnit;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
+import backtype.storm.metric.api.IMetric;
```
Comment (dawillcox) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java]: My IDE automatically cleaned these up for me.

File: pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java [hunk 3]
```diff
@@ -235,7 +233,11 @@ private void mapToValueAndEmit(Message msg) {
                 }
                 ack(msg);
             } else {
-                collector.emit(values, msg);
+                if (values instanceof PulsarTuple) {
```
Comment (dawillcox) [inline pulsar-storm/src/main/java/org/apache/pulsar/storm/PulsarSpout.java]: This right here is the only substantive change to existing code.

**Comment by dawillcox** [comment]
Any guidance on what's needed for this would be appreciated? For example, should I do a separate PR to master?

**Comment by sijie** [comment]
Hi @dawillcox, we are working on the master branch. Can you please send the pull request based on master? 

See the contribution guide: http://pulsar.apache.org/en/contributing/

**Comment by dawillcox** [comment]
OK, new PR created: https://github.com/apache/pulsar/pull/6039
This one presumably should be closed.

**Pull request closed without merging**